### PR TITLE
[Relay] Handle float16 constants & fix BatchNorm

### PR DESCRIPTION
**Pull request by cbalint13**
This PR fixes:

   *  constant expressions targeted as float16 in relay.
   * batchnorm module is not fixed to float32 anymore.
   * extend relay testcase to cover float16 too.

It was tested on real Yolo-Tiny net targeted as float16, it works well now including auto-tuning.

It is based on @anijain2305 suggestion from this [discuss.tvm.ai](https://discuss.tvm.ai/t/relay-nn-batch-norm-fails-with-float16/2536/2) thread.

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks for the contribution, please request reviews from Reviewers

**Comment by cbalint13** [comment]
cc @jroesch @MarisaKirisame @wweic @icemelon9 
  
  Please help with review.

File: src/relay/pass/simplify_inference.cc [hunk 1]
```diff
@@ -52,7 +53,6 @@ Expr BatchNormToInferUnpack(const Attrs attrs,
   }
 
   int axis = param->axis;
-  auto ttype = tdata.as<TensorTypeNode>();
   CHECK(ttype);
```
Comment (vinx13) [inline src/relay/pass/simplify_inference.cc]: let's also move this CHECK to the beginning of the function

File: src/relay/pass/pattern_util.h [hunk 2]
```diff
@@ -204,7 +208,14 @@ template<typename T>
 inline Constant MakeConstantScalar(DataType dtype, T value) {
   runtime::NDArray arr = runtime::NDArray::Empty({}, Type2TVMType(dtype), {kDLCPU, 0});
   TVM_DTYPE_DISPATCH(dtype, DType, {
-    *static_cast<DType*>(arr->data) = value;
+    if (dtype == Float(16)) {
+      // convert to float16
+      // storage is uint16_t
+      *static_cast<DType*>(arr->data) =
+        __truncXfYf2__<float, uint32_t, 23, uint16_t, uint16_t, 10>(static_cast<float>(value));
```
Comment (vinx13) [inline src/relay/pass/pattern_util.h]: please add an assertion that `T` is `float32`

Comment (cbalint13) [reply src/relay/pass/pattern_util.h]: @vinx13 
* static_cast<float>(T value) will make sure __truncXfyf2() is feeding with float32.
* unsure, still want to ```assert(T is_float32)``` or we let T to be any type ?

Comment (anijain2305) [reply src/relay/pass/pattern_util.h]: This bit manipulation seems to bring a certain level of tediousness/ugliness in the TVM codebase.

A way to resolve this can be to keep the Float constants always in FP32. And insert a Relay cast operation to FP16 if need be. Then this will become part of fold_constant in Relay graph, and will use LLVM to generate the FP32 to FP16 conversion (hiding the bit manipulation).

I do not have any strong opinion on either. What do you think @vinx13 ?

Comment (jroesch) [reply src/relay/pass/pattern_util.h]: Because we support arbitrary bit types in the IR, we should really use gmp and mpfr for storing literals. It is bad practice to store them in finite types, as truncation may occur and it is impossible to test whether truncation occurred programmatically w/o weird hacks. This is out of the scope of this PR unfortunately.

Comment (vinx13) [reply src/relay/pass/pattern_util.h]: @cbalint13 ignore my comments :) the static cast if sufficient here

**Comment by vinx13** [comment]
Thanks @cbalint13 @wweic @jroesch @anijain2305  this is merged
